File: backend-express/src/worker/handler/ai.ts
import { Job, Worker } from "bullmq";

import { huggingFace } from "@/lib/clients";
import { crudPlace } from "@/models/crud";

import {
    broker_config,
    PlaceEmbeddingData,
    Queues,
    TASK_PLACE_EMBEDDING,
} from "../config";

// Create Tasks

async function placeEmbeddingTask(job: Job<PlaceEmbeddingData>): Promise<void> {
    const { placeId } = job.data;
    const place = await crudPlace.getDocument(placeId);
    if (!place) {
        console.log(`No place with id ${placeId} found in the database`);
        return;
    }
    const text = `${place.title} - ${place.description}`;
    const result = await huggingFace.embedText(text);
    place.set({ embedding: result });
    crudPlace.saveDocument(place);
    console.log(result);
}

// Tasks Router
type AiTasksData = PlaceEmbeddingData;

async function aiTasksRouter(job: Job<AiTasksData>) {
    switch (job.name) {
        case TASK_PLACE_EMBEDDING:
            await placeEmbeddingTask(job);
            break;
        default:
            throw new Error(`Unknown job name: ${job.name}`);
    }
}

// Start Worker
export const aiWorker = new Worker(Queues.AI, aiTasksRouter, broker_config);
aiWorker.on("failed", (job, err) => {
    console.error(`Job ${job?.id} failed with error ${err.message}`);
});

File: backend-fastapi/background/tasks/ai.py
import logging

from background.setup import TASK_PLACE_EMBEDDING, PlaceEmbbeddingData, Queues
from background.tasks.handler import handler
from services.instances import hf_client


@handler.register_task(TASK_PLACE_EMBEDDING, Queues.AI)
async def place_embedding_task(data: PlaceEmbbeddingData):
    # Lazy loading to avoid circular import issues
    from models.crud import crud_place

    place_id = data["place_id"]
    place = await crud_place.get_document(place_id)
    if place is None:
        logging.warning(f"No place with id {place_id} found in the database")
        return

    text = f"{place.title} - {place.description}"
    result = await hf_client.embed_text(text)
    place.embedding = result
    await crud_place.save_document(place)
    logging.info(result)
